--- a/ath10k-4.19/htt_rx.c
+++ b/ath10k-4.19/htt_rx.c
@@ -2135,7 +2135,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx, wave-1-ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -2229,7 +2228,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx, ack-rssi-filled: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -2246,7 +2244,6 @@ do_generic:
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-4.19/wmi.c
+++ b/ath10k-4.19/wmi.c
@@ -6049,7 +6049,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-4.20/htt_rx.c
+++ b/ath10k-4.20/htt_rx.c
@@ -2189,7 +2189,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -2283,7 +2282,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-4.20/wmi.c
+++ b/ath10k-4.20/wmi.c
@@ -6108,7 +6108,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.10/htt_rx.c
+++ b/ath10k-5.10/htt_rx.c
@@ -3026,7 +3026,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3120,7 +3119,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.10/wmi.c
+++ b/ath10k-5.10/wmi.c
@@ -6329,7 +6329,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.11/htt_rx.c
+++ b/ath10k-5.11/htt_rx.c
@@ -2846,7 +2846,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -2940,7 +2939,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.11/wmi.c
+++ b/ath10k-5.11/wmi.c
@@ -6329,7 +6329,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.12/htt_rx.c
+++ b/ath10k-5.12/htt_rx.c
@@ -3027,7 +3027,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3121,7 +3120,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.12/wmi.c
+++ b/ath10k-5.12/wmi.c
@@ -6329,7 +6329,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.13/htt_rx.c
+++ b/ath10k-5.13/htt_rx.c
@@ -3027,7 +3027,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3121,7 +3120,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.13/wmi.c
+++ b/ath10k-5.13/wmi.c
@@ -6329,7 +6329,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.15/htt_rx.c
+++ b/ath10k-5.15/htt_rx.c
@@ -3025,7 +3025,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3119,7 +3118,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.15/wmi.c
+++ b/ath10k-5.15/wmi.c
@@ -6333,7 +6333,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.17/htt_rx.c
+++ b/ath10k-5.17/htt_rx.c
@@ -3035,7 +3035,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3129,7 +3128,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.17/wmi.c
+++ b/ath10k-5.17/wmi.c
@@ -6358,7 +6358,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.19/htt_rx.c
+++ b/ath10k-5.19/htt_rx.c
@@ -3150,7 +3150,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3244,7 +3243,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.19/wmi.c
+++ b/ath10k-5.19/wmi.c
@@ -6358,7 +6358,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.2/htt_rx.c
+++ b/ath10k-5.2/htt_rx.c
@@ -2411,7 +2411,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -2505,7 +2504,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.2/wmi.c
+++ b/ath10k-5.2/wmi.c
@@ -6224,7 +6224,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.4/htt_rx.c
+++ b/ath10k-5.4/htt_rx.c
@@ -3018,7 +3018,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3112,7 +3111,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.4/wmi.c
+++ b/ath10k-5.4/wmi.c
@@ -6242,7 +6242,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.7/htt_rx.c
+++ b/ath10k-5.7/htt_rx.c
@@ -2831,7 +2831,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -2925,7 +2924,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.7/wmi.c
+++ b/ath10k-5.7/wmi.c
@@ -6294,7 +6294,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-5.8/htt_rx.c
+++ b/ath10k-5.8/htt_rx.c
@@ -2831,7 +2831,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -2925,7 +2924,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-5.8/wmi.c
+++ b/ath10k-5.8/wmi.c
@@ -6323,7 +6323,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-6.2/htt_rx.c
+++ b/ath10k-6.2/htt_rx.c
@@ -3151,7 +3151,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3245,7 +3244,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-6.2/wmi.c
+++ b/ath10k-6.2/wmi.c
@@ -6366,7 +6366,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-6.4/htt_rx.c
+++ b/ath10k-6.4/htt_rx.c
@@ -3151,7 +3151,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3245,7 +3244,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-6.7/htt_rx.c
+++ b/ath10k-6.7/htt_rx.c
@@ -3150,7 +3150,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3244,7 +3243,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-6.9/htt_rx.c
+++ b/ath10k-6.9/htt_rx.c
@@ -3151,7 +3151,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3245,7 +3244,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-6.10/htt_rx.c
+++ b/ath10k-6.10/htt_rx.c
@@ -3151,7 +3151,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3245,7 +3244,6 @@ static void ath10k_htt_rx_tx_compl_ind(s
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
--- a/ath10k-6.4/wmi.c
+++ b/ath10k-6.4/wmi.c
@@ -6358,7 +6358,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-6.7/wmi.c
+++ b/ath10k-6.7/wmi.c
@@ -6366,7 +6366,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-6.9/wmi.c
+++ b/ath10k-6.9/wmi.c
@@ -6387,7 +6387,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
--- a/ath10k-6.10/wmi.c
+++ b/ath10k-6.10/wmi.c
@@ -6387,7 +6387,6 @@ static void ath10k_wmi_event_beacon_tx(s
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
